File: bending-api/src/main/java/pw/dotdash/bending/api/effect/EffectService.java
package pw.dotdash.bending.api.effect;

import com.flowpowered.math.vector.Vector3d;
import org.spongepowered.api.Sponge;
import org.spongepowered.api.effect.particle.ParticleEffect;
import pw.dotdash.bending.api.element.Element;

/**
 * Represents a service used for the creation of particle effects.
 */
public interface EffectService {

    /**
     * Gets the singleton instance of the {@link EffectService}.
     *
     * @return The singleton service instance
     */
    static EffectService getInstance() {
        return Sponge.getServiceManager().provideUnchecked(EffectService.class);
    }

    ParticleEffect getExtinguishEffect();

    ParticleEffect createParticle(Element element, int quantity, Vector3d offset);

    ParticleEffect createRandomParticle(Element element, int quantity);
}

File: src/main/kotlin/arvenwood/bending/api/service/EffectService.kt
package arvenwood.bending.api.service

import arvenwood.bending.api.element.Element
import com.flowpowered.math.vector.Vector3d
import org.spongepowered.api.Sponge
import org.spongepowered.api.effect.particle.ParticleEffect
import org.spongepowered.api.effect.particle.ParticleType

interface EffectService {

    companion object {
        @JvmStatic
        fun get(): EffectService =
            Sponge.getServiceManager().provideUnchecked(EffectService::class.java)
    }

    fun getParticleType(element: Element): ParticleType

    fun createParticle(element: Element, quantity: Int, offset: Vector3d): ParticleEffect

    fun createRandomParticle(element: Element, quantity: Int): ParticleEffect
}